docker-compose.yml

```
@@ -7,10 +6,13 @@ services:
     ports:
       - 3000:3000
     volumes:
-      - ./testCourse:/course
       - /var/run/docker.sock:/var/run/docker.sock
       - ${HOME}/pl_ag_jobs:/jobs
       - .:/PrairieLearn
+      # To test local changes to workspaces or external graders
+      # or override other defaults from ConfigSchema,
+      # enable the following line and make your edits.
+      # - config.dev.json:/PrairieLearn/config.json
```

The conventional way to configure PL is to just drop a `config.json` file into the root of the repository, which will be handled automatically by the `.:/PrairieLearn` mount above.

It's not clear why we're elevating `workspacePullImagesFromDockerHub` and `externalGradingPullImagesFromDockerHub` above all other config options one might want to set. If they're not documented, let's document them alongside other strategies for testing locally-built images and rely on folks setting up their own config file. I'm specifically a fan of a method that doesn't involve any config changes at all: using a specific tag like `prairielearn/grader-c:nathan-testing` that is guaranteed not to exist in the remote image registry.

---

Note for this -- we can autogenerate some config option documentation since we have zod-to-json-schema and jsonschema2md!

---

I don't think there's much value in that. Anyone who is technical enough to want to configure things can and should just read the source code. Most of the useful context about config options (when it exists) is correctly in comments, which won't come across in any kind of JSON schema.

---

Sure, that seems reasonable. Though I do like the idea of pulling out a 'it just works' local config into the documentation, and pointing to the config file on the docs, if we don't already.

---

i.e placing a recommendation for setting these on the external grader/workspace docs or if you work with those files. I have run into the same piece of confusion why my local image changes weren't picked up.

---

This is now tracked in https://github.com/PrairieLearn/PrairieLearn/issues/11934.

---

> It's not clear why we're elevating `workspacePullImagesFromDockerHub` and `externalGradingPullImagesFromDockerHub` above all other config options one might want to set. If they're not documented, let's document them alongside other strategies for testing locally-built images and rely on folks setting up their own config file. I'm specifically a fan of a method that doesn't involve any config changes at all: using a specific tag like `prairielearn/grader-c:nathan-testing` that is guaranteed not to exist in the remote image registry.

I think some of the settings like the two you mentioned may be floating around in several faculty user's own config.json based on informal docs like the one I posted here (which was shared with a few people on Slack also):

https://github.com/PrairieLearn/PrairieLearn/discussions/8897

This might explain why those in particular. I found them useful when I wanted to prefer my local rebuild of a core PL grader/workspace over the upstream one (or to fail fast if I had neglected to build my local one).

---

That's not surprising! They can be useful things to configure in specific situations.

If you have any new relevant context to share, please put it on https://github.com/PrairieLearn/PrairieLearn/issues/11934 where we're now tracking the documentation improvements.